docker/entrypoint.sh

```diff
@@ -6,4 +6,8 @@ set -o pipefail
 
 . /usr/local/share/load-env.sh
 
+if [ -n "$UMASK" ]; then
```

Comment:
This change seems unrelated to the PR itself. Can you please remove it if that's the case?

Reply:
On many UNIX systems, the default umask is 022. 
This significantly limits the usefulness of the new options in the context of "group" and "other" permissions
If we don't use the UMASK environment variable, we have to call umask(2) directly in the go application (hard-coded zero or add new option like -fs-umask=umask).

Reply:
I don't know how umask works. Can you provide some hints or links so I can understand what this is about?

Reply:
Here is an example using Go code
https://stackoverflow.com/a/61645606/23561452

I needed this patch to get group write access to the intermediate directories of the downloaded file.

```
// pre-create hook:
return {
            ChangeFileInfo: {
                ID: "subdir1/subdir2/.../basename",
            },
        };
```

Reply:
Sorry, I meant that I don't know what `umask` is in general.

Reply:
Ok, without this PR, for example, it is not possible to get write/delete access a uploaded file for o(ther) users who are not a tusd user and not in the tusd group, It is also not possible to get write/delete access to созданным subdirectories for members of the tusd group.

Do you know how traditional Unix permissions for files and directories work? man chmod(2)?
Without realizing it, I'm afraid I can't explain it to you succinctly.

Thanks for your work.

Reply:
A simpler solution might have been to use umask(2) with this patch:

```diff
diff --git a/pkg/filestore/filestore.go b/pkg/filestore/filestore.go
index 6f9ab52..c2f5697 100644
--- a/pkg/filestore/filestore.go
+++ b/pkg/filestore/filestore.go

-var defaultFilePerm = os.FileMode(0664)
-var defaultDirectoryPerm = os.FileMode(0754)
+var defaultFilePerm = os.FileMode(0666)
+var defaultDirectoryPerm = os.FileMode(0777)
```

A matter of taste.

Reply:
Ok, so I read up a bit on umask and this is what I understand so far: The umask is a property of the process and it restricts the permissions that are set on newly created files and directories. Even if we tell Go to create a file with permission 0664, the file will actually end up with 0644 if our umask is 022. Is that correct?

If so, this does not seem to be a problem specific to Docker container, but something that affects tusd processes in general.

I like the simpler solution that you are proposing. Instead of defining permissions in the code that won't end up on the created files, we can delegate the control of this to the end user. Either by providing a flag to set the umask on the process or by delegating the work of setting the correct umask to the end user. What do you think? How is this handled in other Docker images?

Reply:
Yes, now you exactly understand the mechanics of how umask(2) works.

```diff
diff --git a/docker/entrypoint.sh b/docker/entrypoint.sh
index 88a70ef..f1934bb 100644
--- a/docker/entrypoint.sh
+++ b/docker/entrypoint.sh
@@ -6,4 +6,18 @@ set -o pipefail
 
 . /usr/local/share/load-env.sh
 
+if printenv UMASK >/dev/null; then
+   umask "$UMASK"
+fi
```

With this and the other above patches and without explicitly setting env UMASK for the docker container, the tusd daemon will create new directories/files with 0755/0644 permissions (since usually linux distributions set umask 022 for users by default).

If a user needs something special, like "wx" access for a group, they will have to pass env UMASK=002 to the container. And then the permissions of newly created directories/files will become 0775/0664.

This changes is lightweight and should satisfy most users. Unless the user wants different permissions for files and directories, e.g. 0755 (g=rx) for directories but 0660 (g=rwx) for files ;)

Reply:
Thanks for the patience. I'm slowly getting it. Do you know how other Docker images handle this? Do they expose a UMASK environment variable as well? Or are there other patterns to control file permissions from containers?

Reply:
The docker ecosystem has "Isolate containers with a user namespace", which would be interesting to use with a TUSD container, but it won't solve the UMASK problem.

Reply:
This doesn't really ask my question, so let me try to rephrase it: Do other Docker images also use the `UMASK` environment variable for configuring the umask of the process running in the Docker container? Does the Docker community have other defacto ways of configuring the umask?

Reply:
Я не использовал никакие другие проекты, где мне нужно было бы менять UMASK. Поэтому мой ответ на ваш вопрос - я не знаю и я не знаю каких либо де-факто решений. 

Однако, если поискать в google, можно найти много постов, где люди ищут подобные решения в других проектах.

Вообще, касательно umask существует не так много решений:
1. pam_umask(8) (systemwide only),
2. RC, like bashrc (systemwide and user),
3. umask (CLI).